julia/GradBench/src/benchmarks/ode.jl

```diff
@@ -12,52 +12,78 @@
 
 module ODE
 
-struct Input
-    x::Vector{Float64}
-    s::Int
-end
+using ADTypes: AbstractADType
+import DifferentiationInterface as DI
 
-function ode_fun(n, x, y, z)
+function ode_fun(x, y, z)
+    n = length(x)
     z[1] = x[1]
     for i in 2:n
         z[i] = x[i] * y[i-1]
     end
 end
 
-function primal(n, xi::Vector{T}, s, yf::Vector{T}) where {T}
+function primal!(y::AbstractVector{T}, x::AbstractVector{T}, s::Real) where {T}
     tf = T(2)
     h = tf / T(s)
 
-    k1 = Vector{T}(undef, n)
-    k2 = Vector{T}(undef, n)
-    k3 = Vector{T}(undef, n)
-    k4 = Vector{T}(undef, n)
-    y_tmp = Vector{T}(undef, n)
+    k1 = similar(y)
+    k2 = similar(y)
+    k3 = similar(y)
+    k4 = similar(y)
+    y_tmp = similar(y)
 
-    yf .= T(0)
+    y .= T(0)
 
     for _ in 1:s
-        ode_fun(n, xi, yf, k1)
+        ode_fun(x, y, k1)
 
-        for i in 1:n
-            y_tmp[i] = yf[i] + h * k1[i] / T(2)
+        for i in eachindex(y_tmp, y, k1)
+            y_tmp[i] = y[i] + h * k1[i] / T(2)
         end
-        ode_fun(n, xi, y_tmp, k2)
+        ode_fun(x, y_tmp, k2)
 
-        for i in 1:n
-            y_tmp[i] = yf[i] + h * k2[i] / T(2)
+        for i in eachindex(y_tmp, y, k2)
+            y_tmp[i] = y[i] + h * k2[i] / T(2)
         end
-        ode_fun(n, xi, y_tmp, k3)
+        ode_fun(x, y_tmp, k3)
 
-        for i in 1:n
-            y_tmp[i] = yf[i] + h * k3[i]
+        for i in eachindex(y_tmp, y, k3)
+            y_tmp[i] = y[i] + h * k3[i]
         end
-        ode_fun(n, xi, y_tmp, k4)
+        ode_fun(x, y_tmp, k4)
 
-        for i in 1:n
-            yf[i] += h * (k1[i] + T(2) * k2[i] + T(2) * k3[i] + k4[i]) / T(6)
+        for i in eachindex(y, k1, k2, k3, k4)
+            y[i] += h * (k1[i] + T(2) * k2[i] + T(2) * k3[i] + k4[i]) / T(6)
         end
     end
+    return nothing
+end
+
+function primal(x::AbstractVector, s::Real)
+    y = similar(x)
+    primal!(y, x, s)
+    return y
+end
+
+function gradientlast(backend::AbstractADType, x::AbstractVector, s::Real)
+    return DI.gradient(last ∘ primal, backend, x, DI.Constant(s))
```

Comment:
This is a different complexity than the C++ implementation.

The C++ benchmark essentially is doing a vJp with an vector, for Enzyme the variant here would be faster ( if activity analysis is successful)

Reply:
I am aware, I went for the easiest implementation first. DI can also support a `pullback` on the last primal component, but for ForwardDiff and ReverseDiff it will be much slower, which is why I didn't consider it at first.

Reply:
Can you add a comment?